### PR TITLE
Run as pod and retry kubectl operations

### PR DESCRIPTION
**Pull request by superseb**
~Draft PR to let QA test first~

https://github.com/rancher/rancher/issues/37553

* Run as job/pod (this resolves the issue of performance when running the script on a remote host and the hard requirement to run it on Linux)
* Retry kubectl in a loop and using `timeout` (this resolves the issue that sometimes a `delete` operation hangs and never returns, while the same `delete` operation again works fine)
* Move cattle data namespaces to cattle namespaces as there are more resources that could block deletion and we need to enhance this some other way (maybe mass deleting certain resources but thats for later)
* Add Dockerfile and drone step to publish `rancher/rancher-cleanup` image that is used in the job (similar to the other images)

To deploy:

```
kubectl create -f rancher-cleanup.yaml
```

To check job logs:
```
kubectl  -n kube-system logs -l job-name=cleanup-job  -f
```

### CODE REVIEW

**Comment by superseb** [comment]
This cannot be merged until the DockerHub repo is created in https://github.com/rancherlabs/eio/issues/984, but it can be reviewed.

**Review by eliyamlevy — APPROVED** [review]

File: rancher-cleanup/cleanup.sh [hunk 1]
```diff
@@ -13,11 +13,18 @@ if [ "$(uname -s)" != "Linux" ]; then
...
 # Check kubectl existence
-if ! which kubectl >/dev/null 2>&1; then
+if ! type kubectl >/dev/null 2>&1; then
   echo "kubectl not found in PATH, make sure kubectl is available"
   exit 1
 fi
 
+# Check timeout existence
+if ! type timeout >/dev/null 2>&1; then
+  echo "kubectl not found in PATH, make sure kubectl is available"
```
Comment (jiaqiluo) [inline rancher-cleanup/cleanup.sh:23]:
```suggestion
  echo "timeout not found in PATH, make sure timeout is available"
```

File: rancher-cleanup/cleanup.sh [hunk 2]
```diff
@@ -28,28 +35,55 @@ echo "=> Printing cluster info for confirmation"
...
 {
   if kubectl get namespace $1; then
-    kubectl get namespace "$1" -o json | tr -d "\n" | sed "s/\"finalizers\": \[[^]]\+\]/\"finalizers\": []/"   | kubectl replace --raw /api/v1/namespaces/$1/finalize -f -
-    kubectl delete --ignore-not-found=true --grace-period=30 namespace $1
+    kcpf namespace $1
+    FINALIZERS=$(kubectl get -o jsonpath="{.spec.finalizers}" namespace "$1")
+    if [ "x${FINALIZERS}" != "x" ]; then
+        echo "Finalizers before for namespace "$1": "${FINALIZERS}""
+        kubectl get -o json namespace "$1" | tr -d "\n" | sed "s/\"finalizers\": \[[^]]\+\]/\"finalizers\": []/"   | kubectl replace --raw /api/v1/namespaces/$1/finalize -f -
+        echo "Finalizers after for namespace "$1": $(kubectl get -o jsonpath="{.spec.finalizers}" namespace "$1")"
```
Comment (jiaqiluo) [inline rancher-cleanup/cleanup.sh:77]: why do we need to patch the finalizer again after calling the `kcpf namespace $1` above?

**Review by jiaqiluo — CHANGES_REQUESTED** [review]

**Comment by jiaqiluo** [comment]
Somehow I get the notification for this PR though it is still a draft.   
Be free to ignore my comments if the PR is not ready for review 😄

File: rancher-cleanup/cleanup.sh (hunk 2 above)
Comment (superseb) [reply rancher-cleanup/cleanup.sh:77]: I've seen both exist on namespaces, `metadata.finalizers` and `spec.finalizers` (where `.spec.finalizers` can't be patched so it needs to be using this construction)

**Review by dkeightley — COMMENTED** [review]
LGTM, tested by QA

**Review by dkeightley — APPROVED** [review]

**Pull request merged**